client/profile/user-context-menu.tsx

```diff
@@ -1,34 +1,79 @@
-import React, { useCallback } from 'react'
+import React, { useCallback, useEffect, useRef, useState } from 'react'
 import styled from 'styled-components'
+import { useRoute } from 'wouter'
+import { ChannelModerationAction } from '../../common/chat'
+import { MULTI_CHANNEL } from '../../common/flags'
 import { SbUserId } from '../../common/users/sb-user'
+import { useSelfUser } from '../auth/state-hooks'
+import { getChatUserProfile, moderateUser } from '../chat/action-creators'
+import { openDialog } from '../dialogs/action-creators'
+import { DialogType } from '../dialogs/dialog-type'
+import { Divider } from '../material/menu/divider'
 import { MenuItem } from '../material/menu/item'
 import { Menu } from '../material/menu/menu'
 import { PopoverProps } from '../material/popover'
 import { inviteToParty, kickPlayer, removePartyInvite } from '../parties/action-creators'
 import { useAppDispatch, useAppSelector } from '../redux-hooks'
+import { openSnackbar } from '../snackbars/action-creators'
 import { colorTextFaint } from '../styles/colors'
+import { subtitle1 } from '../styles/typography'
 import { navigateToWhisper } from '../whispers/action-creators'
 import { navigateToUserProfile } from './action-creators'
 
 const LoadingItem = styled(MenuItem)`
   color: ${colorTextFaint};
 `
 
+const LoadingError = styled.div`
+  ${subtitle1};
+  padding: 16px;
+`
+
 export interface ConnectedUserContextMenuProps {
   userId: SbUserId
   popoverProps: Omit<PopoverProps, 'children'>
 }
 
 export function ConnectedUserContextMenu({ userId, popoverProps }: ConnectedUserContextMenuProps) {
   const dispatch = useAppDispatch()
-  const selfUser = useAppSelector(s => s.auth.user)
+  const selfUser = useSelfUser()
+  const cancelLoadRef = useRef(new AbortController())
+  const [isLoadingChatUserProfile, setIsLoadingChatUserProfile] = useState<boolean>(false)
+  const [loadingChatUserProfileError, setLoadingChatUserProfileError] = useState<Error>()
+  const selfPermissions = useAppSelector(s => s.auth.permissions)
   const user = useAppSelector(s => s.users.byId.get(userId))
+  const [, routeParams] = useRoute('/chat/:channelName')
+  const channelName = routeParams?.channelName.toLowerCase()
+  const channel = useAppSelector(s => (channelName ? s.chat.byName.get(channelName) : undefined))
+  const chatUserProfile = channel?.userProfiles.get(userId)
```

Comment:
This is a bad way to get this data because now this component needs to re-render any time there is a new message in the channel, someone joins or leaves, someone goes online or offline, etc. The ideal way to return this is inside the `useAppSelector` call, as a single thing. I don't think we need the channel object at all so this should be fairly straightforward

Reply:
After commenting above, yes this is also not good to be doing in this component, given that it is a general component and not at all specific to chat.